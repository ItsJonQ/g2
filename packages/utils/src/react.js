import { Children, isValidElement } from 'react';

/**
 * Merges React `ref` together.
 */
export { default as mergeRefs } from 'react-merge-refs';

/**
 * Copies non-react specific statics from a child component to a parent component.
 * Similar to Object.assign, but with React static keywords prevented from being overridden.
 */
export { default as hoistNonReactStatics } from 'hoist-non-react-statics';

/**
 * Gets a collection of available children elements from a React component's children prop.
 *
 * @param {import('react').ReactNode} children
 *
 * @return {import('react').ReactNodeArray} An array of available children.
 */
export function getValidChildren(children) {
	if (typeof children === 'string') return [children];

	return Children.toArray(children).filter((child) => isValidElement(child));
}

/**
 * Gets the displayName of a React Component or element.
 *
 * @param {string | import('react').ComponentType} tagName
 *
 * @return {string} The display name of the Component / tagName.
 */
export function getDisplayName(tagName) {
	const displayName =
		typeof tagName === 'string'
			? tagName
			: tagName.displayName || tagName.name || 'Component';

	return displayName;
}

/**
 * Checks to see if a `children` prop is a render function prop.
 *
 * @param {*} children
 *
 * @return {boolean} True, if children is a render function prop.
 */
export function isRenderProp(children) {
	return typeof children === 'function';
}

/**
 * Handles the rendering of a React component's children prop, which
 * may be a render function prop.
 *
 * @param {*} children Children to render.
 * @param {*} props Props to pass into a (potential) children render function.
 *
 * @return {*} The rendered children.
 */
export function renderChildren(children, props = {}) {
	if (isRenderProp(children)) {
		// eslint-disable-next-line no-unused-vars
		const { children: _, ...rest } = props;
		return children(rest);
	}

	return children;
}
