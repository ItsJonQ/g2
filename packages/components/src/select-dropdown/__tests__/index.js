import { render } from '@testing-library/react';
import React from 'react';

import { SelectDropdown } from '../index';

describe('props', () => {
	test('should render correctly', () => {
		const { container } = render(<SelectDropdown />);
		expect(container.firstChild).toMatchSnapshot();
	});
});
