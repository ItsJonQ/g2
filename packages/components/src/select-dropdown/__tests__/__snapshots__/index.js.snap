// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render correctly 1`] = `
.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  position: relative;
}

@media (prefers-reduced-motion) {
  .emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0.emotion-0 {
  -webkit-transition: none!important;
  transition: none!important;
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  --wp-g2-flex-gap: calc(4px * 0);
  --wp-g2-flex-gap: calc(var(--wp-g2-grid-base) * 0);
  --wp-g2-flex-item-margin-bottom: 0;
  --wp-g2-flex-item-margin-right: calc(4px * 2);
  --wp-g2-flex-item-margin-right: var(--wp-g2-flex-gap);
  --wp-g2-flex-item-margin-left: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  background: #ffffff;
  background: var(--wp-g2-control-background-color);
  border-radius: 2px;
  border-radius: var(--wp-g2-control-border-radius);
  border: 1px solid;
  border-color: #757575;
  border-color: var(--wp-g2-control-border-color);
  box-shadow: transparent;
  box-shadow: var(--wp-g2-control-box-shadow);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  outline: none;
  padding: 0 8px;
  position: relative;
  -webkit-transition: border-color 100ms ease;
  -webkit-transition: border-color var(--wp-g2-transition-duration-fastest) ease;
  transition: border-color 100ms ease;
  transition: border-color var(--wp-g2-transition-duration-fastest) ease;
  width: 100%;
  cursor: pointer;
  padding: 0;
}

@media (prefers-reduced-motion) {
  .emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1 {
  -webkit-transition: none!important;
  transition: none!important;
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1>*+*:not(marquee) {
  margin-left: calc(4px * 0);
  margin-left: calc(var(--wp-g2-grid-base) * 0);
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1>* {
  min-width: 0;
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1[disabled] {
  opacity: 0.6;
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1:hover {
  border-color: #757575;
  border-color: var(--wp-g2-control-border-color-hover);
}

.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1:focus,
.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1.emotion-1[data-focused='true'] {
  border-color: #007cba;
  border-color: var(--wp-g2-color-admin);
  box-shadow: 0 0 0 0.5px #007cba;
  box-shadow: var(--wp-g2-control-box-shadow-focus);
}

.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: 2px;
  border-radius: var(--wp-g2-control-border-radius);
  box-shadow: none;
  box-sizing: border-box;
  color: #1e1e1e;
  color: var(--wp-g2-color-text);
  display: block;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  line-height: 18px;
  min-height: calc(30px - 2px);
  min-height: calc(var(--wp-g2-control-height) - 2px);
  outline: none;
  padding: 0;
  padding-bottom: calc((30px - 2px - 18px) / 2);
  padding-bottom: calc((var(--wp-g2-control-height) - 2px - 18px) / 2);
  padding-top: calc((30px - 2px - 18px) / 2);
  padding-top: calc((var(--wp-g2-control-height) - 2px - 18px) / 2);
  resize: none;
  text-align: left;
  width: 100%;
  font-size: 16px;
  font-size: var(--wp-g2-font-size-input-mobile);
  padding-left: 8px;
  padding-right: 30px;
  padding-right: 24px;
}

@media (prefers-reduced-motion) {
  .emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2 {
  -webkit-transition: none!important;
  transition: none!important;
}

@media (min-width:36em) {
  .emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2 {
    font-size: 13px;
    font-size: var(--wp-g2-font-size);
  }
}

.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2::-webkit-outer-spin-button,
.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2.emotion-2[type='number'] {
  -moz-appearance: textfield;
}

.emotion-3.emotion-3.emotion-3.emotion-3.emotion-3.emotion-3.emotion-3 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

@media (prefers-reduced-motion) {
  .emotion-3.emotion-3.emotion-3.emotion-3.emotion-3.emotion-3.emotion-3 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-3.emotion-3.emotion-3.emotion-3.emotion-3.emotion-3.emotion-3 {
  -webkit-transition: none!important;
  transition: none!important;
}

.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  display: block;
  max-height: 100%;
  max-width: 100%;
  min-height: 0;
  min-width: 0;
  display: var(--wp-g2-flex-item-display);
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0 5px;
  pointer-events: none;
  position: absolute;
  right: 0;
  top: 0;
}

@media (prefers-reduced-motion) {
  .emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4 {
  -webkit-transition: none!important;
  transition: none!important;
}

@media (prefers-reduced-motion) {
  .emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4.emotion-4 {
  -webkit-transition: none!important;
  transition: none!important;
}

.emotion-5.emotion-5.emotion-5.emotion-5.emotion-5.emotion-5.emotion-5 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  color: #1e1e1e;
  color: var(--wp-g2-color-text);
  line-height: 1.2;
  line-height: var(--wp-g2-font-line-height-base);
  font-size: calc((13 / 13) * 13px);
  font-size: calc((13 / 13) * var(--wp-g2-font-size));
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  display: block;
  color: #717171;
  color: var(--wp-g2-color-text-muted);
}

@media (prefers-reduced-motion) {
  .emotion-5.emotion-5.emotion-5.emotion-5.emotion-5.emotion-5.emotion-5 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-5.emotion-5.emotion-5.emotion-5.emotion-5.emotion-5.emotion-5 {
  -webkit-transition: none!important;
  transition: none!important;
}

.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
  display: block;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  color: currentColor;
  fill: currentColor;
  height: 16px;
  width: 16px;
}

@media (prefers-reduced-motion) {
  .emotion-6.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-6.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6 {
  -webkit-transition: none!important;
  transition: none!important;
}

.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6.emotion-6 svg {
  display: block;
}

.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7 {
  box-sizing: border-box;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-family: Inter,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",sans-serif;
  font-family: var(--wp-g2-font-family);
  font-size: 13px;
  font-size: var(--wp-g2-font-size);
  font-weight: normal;
  font-weight: var(--wp-g2-font-weight);
  margin: 0;
}

@media (prefers-reduced-motion) {
  .emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7 {
    -webkit-transition: none!important;
    transition: none!important;
  }
}

[data-system-ui-reduced-motion-mode="true"] .emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7.emotion-7 {
  -webkit-transition: none!important;
  transition: none!important;
}

<div
  class="components-select-dropdown wp-components-select-dropdown ic-69e0bf ic-192yz5d emotion-0"
  data-g2-c16t="true"
  data-g2-component="SelectDropdown"
>
  <iframe
    aria-hidden="true"
    frameborder="0"
    src="about:blank"
    style="display: block; opacity: 0; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; overflow: hidden; pointer-events: none; z-index: -1;"
    tabindex="-1"
  />
  <div
    class="components-flex wp-components-flex components-base-field wp-components-base-field ic-172ogxm ic-17cs4ki components-select wp-components-select ic-gutgg1 ic-192yz5d emotion-1"
    data-g2-c16t="true"
    data-g2-component="SelectWrapper"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="ic-192yz5d emotion-2"
      data-g2-c16t="true"
      data-g2-component="SelectDropdownReference"
      value=""
    >
      <span
        class="components-truncate wp-components-truncate ic-1m45iq3 ic-192yz5d emotion-3"
        data-g2-c16t="true"
        data-g2-component="Truncate"
      />
    </button>
    <div
      class="components-flex-item wp-components-flex-item ic-lqthxd ic-192yz5d ic-1p3tlc8 ic-192yz5d emotion-4"
      data-g2-c16t="true"
      data-g2-component="FlexItem"
    >
      <span
        class="components-truncate wp-components-truncate components-text wp-components-text ic-7u3orf ic-1m45iq3 ic-192yz5d emotion-5"
        data-g2-c16t="true"
        data-g2-component="Text"
      >
        <div
          class="components-icon wp-components-icon ic-2q44rw ic-192yz5d emotion-6"
          data-g2-c16t="true"
          data-g2-component="Icon"
        >
          <svg
            aria-hidden="true"
            focusable="false"
            height="16"
            role="img"
            size="16"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17.5 11.6L12 16l-5.5-4.4.9-1.2L12 14l4.5-3.6 1 1.2z"
            />
          </svg>
        </div>
      </span>
    </div>
  </div>
  <label
    class="ic-192yz5d emotion-7"
    data-g2-component="SelectDropdownLabel"
    for="downshift-0-toggle-button"
    id="downshift-0-label"
  />
</div>
`;
